src/vset.h

```diff
@@ -0,0 +1,312 @@
+ *     VSET_BUCKET_SINGLE  - A single pointer to one entry.
+ *     VSET_BUCKET_VECTOR  - A sorted vector of pointers (up to 127 entries).
+ *     VSET_BUCKET_HT      - A hashtable used when vectors become too dense.
+ *
+ * Bucket Timestamp (END of window):
+ *
+ *        |------------------ Bucket Span ------------------|
+ *        [window_start .................................. bucket_ts)
+ *
+ * ASCII Layout Example:
+ *
+ *   Timeline:             ---> increasing time --->
+ *                         +------+---------+--------+
+ *                         | B0   |   B1    |   B2   |
+ *                         | ts=32| ts=128  | ts=2048|
+ *                         +------+---------+--------+
+ *                         ^      ^         ^
+ *                         |      |         |
+ *               [E1,E2] ∈ B0   [E3...E7] ∈ B1    [E8...] ∈ B2
+ *               All entries expire BEFORE their bucket_ts
+ *
+ * Bucket Splitting Strategy:
+ * ----------------------------------
+ *
+ * When a bucket (e.g. VECTOR) becomes too dense or needs realignment:
+ *
+ * 1. **Re-align to lower granularity:**
+ *      - Adjust the bucket timestamp down to a finer granularity (e.g. 16ms).
+ *      - Only done if ALL entries still fit in the tighter window.
+ *      - Effectively “moves” the bucket to an earlier timestamp.
+ *
+ *        Example: B(ts=128, span=128ms) -> B(ts=64, span=16ms)
+ *
+ * 2. **Split into two buckets:**
+ *      - Use binary search to find a “natural” boundary based on entry expiry.
+ *      - Original bucket retains its timestamp (but holds fewer entries).
+ *      - New bucket is inserted before the current one with its own tighter timestamp.
+ *
+ *        Example:
+ *
+ *        Before:
+ *             [ Entry0 ... Entry126 ]  -> B(ts=128)
+ *
+ *        After Split:
+ *             [ Entry0...Entry62 ]     -> New B(ts=64)
+ *             [ Entry63...Entry126 ]   -> Original B(ts=128)
+ *
+ * 3. **Convert to hashtable:**
+ *      - When no clean split is found (e.g. all entries share similar expiry),
+ *        and realignment is not possible.
+ *      - This allows efficient O(1) lookups even with clustered expiry values.
+ *
+ *        Vector B(ts=128) -> Hashtable B(ts=128)
+ *
+ * This hierarchical design ensures:
+ *   - Efficient memory usage (tight buckets)
+ *   - Predictable iteration by expiry time
+ *   - Low overhead insertions & deletions
+ *   - Graceful promotion & demotion of bucket types
+ *
+ * NOTE: Buckets are always sorted by their `bucket_ts` in the radix tree (RAX),
+ *       which allows efficient search for insertion/removal based on expiry.
+ *
+ *-----------------------------------------------------------------------------
+ * RAX Bucket Layout
+ *-----------------------------------------------------------------------------
+ *
+ * * RAX View with Time Keys:
+ *
+ *     expiry_buckets = rax * | 0x6
+ *
+ *     +--------------------------+
+ *     | RAX (key = bucket_ts)   |
```

Comment:
The right side is not straight.

```suggestion
 *     | RAX (key = bucket_ts)    |
```